### PR TITLE
School Library:Feature/set up associations

### PR DESCRIPTION
**Pull request by Megagig**
Dear Code Reviewer,
I would like to have a review of my code to be able to merge it.
Here is a little summary of what I have done:

[General requirements](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/school_library_associations.md#general-requirements)
Make sure that there are [no linter errors](https://github.com/microverseinc/linters-config).
Make sure that you used correct [Gitflow](https://github.com/microverseinc/curriculum-transversal-skills/blob/main/git-github/articles/gitflow.md).
Make sure that you documented your work [in a professional way](https://github.com/microverseinc/curriculum-transversal-skills/blob/main/documentation/articles/professional_repo_rules.md).

[Ruby requirements](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/school_library_associations.md#ruby-requirements)
Follow our list of [best practices for Ruby](https://github.com/microverseinc/curriculum-ruby/blob/main/articles/ruby_best_practices.md).

[Project requirements](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/school_library_associations.md#project-requirements)
Create a class Classroom with the following:

@label instance variable, should be initialized in the constructor.

Setter and getter for @label (remember about attr_accessor).
Create the has-many/belongs-to relationship between Classroom and Student. The following should be implemented:
Create the has-many side (a classroom has many students).
Create the belongs-to side (a student belongs to a classroom).
Make sure that when adding a student to a classroom it also sets the classroom for the student.
Make sure that when setting the classroom for a student it also adds it to the classrooms' students.

Create a class Book with the following:

@title and @author instance variables, should be initialized in the constructor.
Setters and getters for instance variables (remember about attr_accessor).

Create a class Rental with the following:

@date instance variable, should be initialized in the constructor.
Setter and getter for @date (remember about attr_accessor).

Create the many-to-many (also has-many-through) relationship between Person and Book using the intermediate class Rental. The following should be implemented:

Create the has-many side of Book and Rental (a book has many rentals).
Create the belongs-to side of Rental and Book (a rental belongs to a book).
Create the has-many side of Person and Rental (a person has many rentals).
Create the belongs-to side of Rental and Person (a rental belongs to a person).
Modify the constructor of Rental so Book and Person are set in it.

### CODE REVIEW

File: student_class.rb [hunk 1]
```diff
@@ -6,6 +6,7 @@ class Student < Person
   def initialize(classroom, age, name = 'unknown', parent_permission: true)
     super(age, name, parent_permission: parent_permission)
     @classroom = classroom
+    classroom&.add_student(self)
   end
```
Comment (1GORDON) [inline student_class.rb]: Great job @Megagig 🤝
- As stated in the project requirements, kindly ensure there is a method that serves as the 'belongs-to' side (a student belongs to a classroom). You may use this [link](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/articles/oop_relationships_by_examples.md#both-ways-relationships) for more clarity.
belongs-to-classroom

File: book.rb [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+  def initialize(title, author)
+    @title = title
+    @author = author
+    @rentals = []
+  end
+
+  def add_rental(rental)
+    @rentals.push(rental)
+    @rentals << rental
+  end
```
Comment (1GORDON) [inline book.rb]: Great job @Megagig 🤝
- As stated in the project requirements, For the add_rental method, kindly ensure the book itself should be sent as a parameter to create the new rental achieving the 'has-many' association.
add-rental-book

File: person_class.rb [hunk 3]
```diff
@@ -11,8 +11,13 @@ def initialize(name = 'unknown', age = nil, parent_permission: true)
     @name = name
     @age = age
     @parent_permission = parent_permission
+    @rentals = []
   end
 
+  def add_rental(rental)
+    @rentals.push(rental)
+    @rentals << rental
+  end
```
Comment (1GORDON) [inline person_class.rb]: Great job @Megagig 🤝

- As stated in the project requirements, For the add_rental method, kindly ensure the person itself should be sent as a parameter to create the new rental achieving the 'has-many' association.
add-rental-person

**Review by 1GORDON — CHANGES_REQUESTED** [review]
##
## Hi @Megagig  ✋

Good job so far!

## Highlights!
- Created the `many-to-many`  relationships✔️ 
- Created `belongs-to` relationships ✔️

Hello ✋, you have done well reaching this point of the milestone 🤝 . But there are some issues that you still need to work on to go to the next project but you are almost there!

# Status: Required Changes ♻️⛔️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend considering them as they can improve your code._

Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread and tag me `@1GORDON` if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: student_class.rb (hunk 1 above)
Comment (Megagig) [reply student_class.rb]: @1GORDON Thank you so much for such a thorough review. This has been corrected.

File: book.rb (hunk 2 above)
Comment (Megagig) [reply book.rb]: @1GORDON Thanks

File: person_class.rb (hunk 3 above)
Comment (Megagig) [reply person_class.rb]: @1GORDON Thanks for reminding me of this requirement.

**Review by KayLemba — APPROVED** [review]
# STATUS: APPROVED 🟢 
Hello @Megagig   ,
Your project is complete! There is nothing else to say other than... it's time to merge it  .
Congratulations! 🎉 Keep up the Awesome work 🥂 Keep Rocking 🥇

HIGHLIGHTS:

- [x]  Your repo is professional 😎 

- [x] Descriptive P.R 💯 Professional README file

- [x] App runs as intended

- [x] Association set up correctly

Now you can merge this branch and move on 🥇
Happy Coding cheers 🥇 💯

Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better.

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using

**Pull request merged**